pkg/sdn/plugin/pod.go

```diff
@@ -0,0 +1,412 @@
+package plugin
+
+import (
+	"encoding/json"
+	"fmt"
+	"net"
+	"os/exec"
+	"strconv"
+	"strings"
+	"syscall"
+
+	"github.com/openshift/origin/pkg/util/netutils"
+
+	kapi "k8s.io/kubernetes/pkg/api"
+	knetwork "k8s.io/kubernetes/pkg/kubelet/network"
+	kbandwidth "k8s.io/kubernetes/pkg/util/bandwidth"
+
+	"github.com/containernetworking/cni/pkg/invoke"
+	"github.com/containernetworking/cni/pkg/ip"
+	"github.com/containernetworking/cni/pkg/ipam"
+	"github.com/containernetworking/cni/pkg/ns"
+	cnitypes "github.com/containernetworking/cni/pkg/types"
+
+	"github.com/vishvananda/netlink"
+
+	// Must be imported to register API types like ClusterNetwork
+	_ "github.com/openshift/origin/pkg/api/install"
+)
+
+const (
+	sdnScript   = "openshift-sdn-ovs"
+	setUpCmd    = "setup"
+	tearDownCmd = "teardown"
+	statusCmd   = "status"
+	updateCmd   = "update"
+
+	AssignMacVlanAnnotation string = "pod.network.openshift.io/assign-macvlan"
+
+	podInterfaceName = knetwork.DefaultInterfaceName
+)
+
+type PodConfig struct {
+	vnid             string
+	ingressBandwidth string
+	egressBandwidth  string
+	wantMacvlan      bool
+}
+
+func getBandwidth(pod *kapi.Pod) (string, string, error) {
+	ingress, egress, err := kbandwidth.ExtractPodBandwidthResources(pod.Annotations)
+	if err != nil {
+		return "", "", fmt.Errorf("failed to parse pod bandwidth: %v", err)
+	}
+	var ingressStr, egressStr string
+	if ingress != nil {
+		ingressStr = fmt.Sprintf("%d", ingress.Value())
+	}
+	if egress != nil {
+		egressStr = fmt.Sprintf("%d", egress.Value())
+	}
+	return ingressStr, egressStr, nil
+}
+
+func wantsMacvlan(pod *kapi.Pod) (bool, error) {
+	privileged := false
+	for _, container := range pod.Spec.Containers {
+		if container.SecurityContext.Privileged != nil && *container.SecurityContext.Privileged {
+			privileged = true
+			break
+		}
+	}
+
+	val, ok := pod.Annotations[AssignMacVlanAnnotation]
+	if !ok || val != "true" {
+		return false, nil
+	}
+	if !privileged {
+		return false, fmt.Errorf("pod has %q annotation but is not privileged", AssignMacVlanAnnotation)
+	}
+
+	return true, nil
+}
+
+func (node *OsdnNode) getPodConfig(req *podRequest) (*PodConfig, error) {
+	config := &PodConfig{}
+
+	if node.multitenant {
+		vnid, err := node.vnids.GetVNID(req.podNamespace)
+		if err != nil {
+			return nil, err
+		}
+		config.vnid = strconv.FormatUint(uint64(vnid), 10)
+	}
+
+	pod, err := node.kClient.Pods(req.podNamespace).Get(req.podName)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read pod %s/%s: %v", req.podNamespace, req.podName, err)
+	}
+
+	config.wantMacvlan, err = wantsMacvlan(pod)
+	if err != nil {
+		return nil, err
+	}
+
+	config.ingressBandwidth, config.egressBandwidth, err = getBandwidth(pod)
+	if err != nil {
+		return nil, err
+	}
+
+	return config, nil
+}
+
+// Returns host veth, container veth MAC, and pod IP
+func getVethInfo(netns, containerIfname string) (netlink.Link, string, string, error) {
```

Comment:
I'd commented before about how the comment claimed it returned the veth name but it actually returned the whole netlink.Link, and you changed the comment, but I'd meant to change the return value, because all of its callers only use the name, not anything else